Fix #63. Add --input-meta parameter to explicitly specify the jsonl field dtypes

miguelusque opened the pull request:
When reading jsonl files with Dask, the dataframe datatypes are inferred unless explicitly specified.

Inferring the data types can lead to several issues, such as incorrect type inference, degradation of performance and increased memory usage among others.

I think we could mitigate those issues if we would add a `--input-meta` parameter, which would receive a dictionary of datatypes.

That parameter would be optional, and be similar to the `--meta` parameter available here: https://docs.dask.org/en/latest/generated/dask.dataframe.read_json.html.

ryantwolf commented on nemo_curator/download/doc_builder.py:
```diff
@@ -190,8 +193,9 @@ def download_and_extract(
         output_type=output_type,
         keep_raw_download=keep_raw_download,
-        force_download=force_download,
+        force_download=force_dssownload,
```
Nit, fix typo

miguelusque replied:
Done!

ryantwolf commented on examples/blend_and_shuffle.py:
```diff
@@ -46,6 +49,14 @@ def attach_args(
         formatter_class=argparse.ArgumentDefaultsHelpFormatter
     ),
 ):
+    parser.add_argument(
```
I would not add this argument to the scripts in `examples/`. The examples are meant to be as simple as possible. The only reason why we have the argparse in the first place here is so that the examples can be easily run on SLURM without modifications.

ryantwolf commented on nemo_curator/scripts/download_and_extract.py:
```diff
@@ -116,6 +117,13 @@ def attach_args(
         help="Path to input data directory",
     )
+    parser.add_argument(
+        "--input-meta",
```
You should add this argument to every script in this folder that deals with a single dataset as input.

miguelusque replied:
I tried to do it to any script that invoked (directly or indirectly) any of the methods where `input_meta` is used. Please, let me know if I have missed anything. Thanks!

ryantwolf commented on nemo_curator/utils/distributed_utils.py:
```diff
@@ -186,6 +187,8 @@ def read_single_partition(
         files: The path to the jsonl files to read.
         backend: The backend to use for reading the data. Either "cudf" or "pandas".
         add_filename: Whether to add a "filename" column to the DataFrame.
+        input-meta: A dictionary with the json object field names and data types.
```
It would be good if this parameter could be a `Union[str, dict]` so that users could directly pass in a dictionary if they are using the Python API, or use the `str` if they are on the command line. Please update the docstring when you do this too to mention that the string should be "a string representation of a dictionary" or something along those lines.

miguelusque replied:
Done!

ryantwolf commented on tutorials/peft-curation/main.py:
```diff
@@ -163,7 +165,15 @@ def run_curation_pipeline(args: Any, jsonl_fp: str) -> str:
     parser = argparse.ArgumentParser()
     parser = add_distributed_args(parser)
+    parser.add_argument(
+        "--input-meta",
```
Since we haven't run into any issues in the tutorial datasets, and for the clarity of the tutorials, I think you should omit the argument from them.

ryantwolf reviewed (commented):
I like this idea! I left a few changes I'd like to see made, but nothing major. Also it looks like there are some merge conflicts so you should probably resolve those.

ryantwolf commented:
After sleeping on it, I do have one last nit. I think the argument for all the `read_*` functions should be `meta`. The argparse arguments can still be `--input-meta`. I want to try and keep the arguments and behavior as close to Dask as possible.

miguelusque commented:
> After sleeping on it, I do have one more nit. I think the argument for all the `read_*` functions should be `meta`. The argparse arguments can still be `--input-meta`. I want to try and keep the arguments and behavior as close to Dask as possible.

Hi @ryantwolf , thank you for all your feedback.

I am addressing all the comments you have made. My first thought about the name of the parameter was to name it `meta`, but then I thought that, in the near future, we may want to also specify the `output-meta` parameter, where we may want to specify specific formats for the output files.

Even if I think it is very likely that `meta` might work for both use-cases, I thought it would be clearer to differentiate them.

Looking forward to your thoughts. Thanks!!!

ayushdg commented on nemo_curator/utils/distributed_utils.py:
```diff
@@ -218,6 +226,11 @@ def read_single_partition(
     else:
         raise RuntimeError("Could not read data, please check file type")
 
+    if input_meta:
```
Same recommendation as above

miguelusque replied:
Done!

ayushdg commented on nemo_curator/utils/distributed_utils.py:
```diff
@@ -218,6 +226,11 @@ def read_single_partition(
     else:
         raise RuntimeError("Could not read data, please check file type")
 
+    if input_meta:
```
Since this argument is only valid for `jsonl` formats, would it make sense to move it to the section above where `filetype=="jsonl"`? We can raise a warning if the user specifies this for other filetypes saying it'll not do anything.

miguelusque replied:
Great comment. Will do it right now.

ayushdg commented on nemo_curator/scripts/verify_classification_results.py:
```diff
@@ -136,9 +148,16 @@ def verify_results(
     """
-    expected_df = pd.read_json(expected_results_file_path, lines=True)
+    if input_meta:
+        read_kwargs["dtype"] = (
```
This file doesn't use `read_kwargs` perhaps you meant to update `input_meta` here?

miguelusque replied:
Great catch! Thank you!

I have actually changed a bit the condition. Because the interface expects either a dictionary, a string or None, I only the string needs to be converted to dictionary, I have changed the logic a bit as follows:

> if type(input_meta) == str:
>     input_meta = ast.literal_eval(input_meta)

ayushdg commented on nemo_curator/utils/fuzzy_dedup_utils/io_utils.py:
```diff
@@ -26,31 +27,41 @@
     of the input file.
     """
+    if input_meta:
+        read_kwargs["dtype"] = (
```
This file also doesn't use `read_kwargs`, probably want to update `input_meta`

miguelusque replied:
Great catch! Thank you!

I have actually changed a bit the condition. Because the interface expects either a dictionary, a string or None, I only the string needs to be converted to dictionary, I have changed the logic a bit as follows:

>     if type(input_meta) == str:
>         input_meta = ast.literal_eval(input_meta)

ayushdg reviewed (changes requested):
Thanks for the changes! Left a few suggestions, but generally looks good. 

In addition to the changes, is it also possible to add a couple of tests, that attempt to use `DocumentDataset.read_json` with the `input_meta` param set both as a string and a dict, and verify that the result dtypes are what we'd expect.

miguelusque commented on nemo_curator/scripts/verify_classification_results.py:
```diff
@@ -136,9 +148,16 @@ def verify_results(
-    expected_df = pd.read_json(expected_results_file_path, lines=True)
+    if input_meta:
+        read_kwargs["dtype"] = (
+            ast.literal_eval(input_meta) if type(input_meta) == str else input_meta
```
Hi Ayush, you are right. The rationale behind accepting strings is to allow passing them from the CLI.

miguelusque commented on nemo_curator/scripts/verify_classification_results.py:
```diff
@@ -136,9 +148,16 @@ def verify_results(
 
     """
-    expected_df = pd.read_json(expected_results_file_path, lines=True)
+    if input_meta:
```
Done!

miguelusque commented:
> Thanks for the changes! Left a few suggestions, but generally looks good.
> 
> In addition to the changes, is it also possible to add a couple of tests, that attempt to use `DocumentDataset.read_json` with the `input_meta` param set both as a string and a dict, and verify that the result dtypes are what we'd expect.

I have just added a test_io.py file to start incorporating IO tests in that file.

ayushdg reviewed (approved):
Thanks a lot for ddressing all my comments.
